### PR TITLE
Add an example of Segment Anything Model [Inference]

### PR DESCRIPTION
**Pull request by cfgfung**
# What does this PR do?

Add an example code of SegmentAnythingModel using Graph mode and BF16.

Original FP32:
n_iterations: 20
Total latency (ms): 10380.430936813354
Average latency (ms): 519.0215468406677

Enabled graph mode and BF16:
n_iterations: 20
Total latency (ms): 1639.5680904388428
Average latency (ms): 81.97840452194214

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by cfgfung** [comment]
@regisss 
This is a validated model and this is an example script for users. No core code has been changed.

File: examples/object-segementation/ClipSeg/run_example.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        texts.append(text)
+
+    if args.use_hpu_graphs:
+        model = ht.hpu.wrap_in_hpu_graph(model)
+
+    autocast = torch.autocast(device_type="hpu", dtype=torch.bfloat16, enabled=args.bf16)
+    model.to("hpu")
+
+    with torch.no_grad(), autocast:
+        for i in range(args.warmup):
```
Comment (yeonsily) [inline examples/object-segementation/ClipSeg/run_example.py:88]: can we change this warmup step checking similar to https://github.com/huggingface/optimum-habana/blob/main/optimum/habana/transformers/trainer.py#L881?

File: examples/object-segementation/SegmentAnythingModel/run_example.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    points = [[points]]
+
+    if args.use_hpu_graphs:
+        model = ht.hpu.wrap_in_hpu_graph(model)
+
+    autocast = torch.autocast(device_type="hpu", dtype=torch.bfloat16, enabled=args.bf16)
+    model.to("hpu")
+
+    with torch.no_grad(), autocast:
+        for i in range(args.warmup):
```
Comment (yeonsily) [inline examples/object-segementation/SegmentAnythingModel/run_example.py:88]: same here.

**Comment by yeonsily** [comment]
@cfgfung Can you please provide GPU benchmark and add ci tests?

**Comment by cfgfung** [comment]
> @cfgfung Can you please provide GPU benchmark and add ci tests?

@yeonsily 

Sure. Is there any reference for the CI test? This is just a validated example. I checked another merged example  (https://github.com/huggingface/optimum-habana/commit/2f55de3f9ad23a853768c16029c455ee11b0e515)  and that does not require CI.

File: examples/object-segementation/ClipSeg/run_example.py (hunk 1 above)
Comment (cfgfung) [reply examples/object-segementation/ClipSeg/run_example.py:88]: That may not apply to this case. The code in the link is for 'training'. There are some variables dedicated for training.  
For example: _args.gradient_accumulation_steps_ , _epoch * steps_in_epoch_ 
These should not be included in this 'inference' example.

File: examples/object-segementation/SegmentAnythingModel/run_example.py (hunk 2 above)
Comment (cfgfung) [reply examples/object-segementation/SegmentAnythingModel/run_example.py:88]: Replied above.

**Comment by jiminha** [comment]
@cfgfung  Your code looks good. Can you just modify your test file name to something like test_image_segementation.py . Also, please make sure your test runs on CI test. 
https://github.com/huggingface/optimum-habana/blob/8786b7592c58d394f9460710415de3c08775b1b6/Makefile#L45

**Comment by cfgfung** [comment]
> @cfgfung Your code looks good. Can you just modify your test file name to something like test_image_segementation.py . Also, please make sure your test runs on CI test.
> 
> https://github.com/huggingface/optimum-habana/blob/8786b7592c58d394f9460710415de3c08775b1b6/Makefile#L45

Thanks for the review. I have renamed the test file and add few lines of codes to the makefile for the CI tests.

File: examples/object-segementation/SegmentAnythingModel/run_example.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    parser.add_argument("--warmup", type=int, default=3, help="Number of warmup iterations for benchmarking.")
+    parser.add_argument("--n_iterations", type=int, default=5, help="Number of inference iterations for benchmarking.")
+
+    args = parser.parse_args()
+
+    adapt_transformers_to_gaudi()
+
+    processor = SamProcessor.from_pretrained(args.model_name_or_path)
+    model = SamModel.from_pretrained(args.model_name_or_path)
+
```
Comment (jiminha) [inline examples/object-segementation/SegmentAnythingModel/run_example.py:74]: Instead have this code specific to SAM model, can you use AutoProcessor, AutoModel and have this script generic?

Comment (jiminha) [reply examples/object-segementation/SegmentAnythingModel/run_example.py:74]: Also, plz remove this specific SAM folder(we don't have specific model folder in all examples), and move the script under object-segmentation/

**Comment by cfgfung** [comment]
> Could you delete old file : test_modelenabling.py?
> 
> Also I ran your tests on the latest, and I'm getting this error on the last test. Do you see this passing on Docker 1.15.0? tests/test_image_segmentation.py::GaudiSAMTester::test_inference_default PASSED tests/test_image_segmentation.py::GaudiSAMTester::test_inference_hpu_graphs PASSED tests/test_image_segmentation.py::GaudiSAMTester::test_no_latency_regression_bf16 FAILED
> 
> FAILED tests/test_image_segmentation.py::GaudiSAMTester::test_no_latency_regression_bf16 - AssertionError: 81.58655166625977 not greater than or equal to 93.97604942321777

Removed the file.
Somehow it passed on my side. I fixed the logic of that test. It should be able to pass now.

File: examples/object-segementation/SegmentAnythingModel/run_example.py (hunk 3 above)
Comment (cfgfung) [reply examples/object-segementation/SegmentAnythingModel/run_example.py]: Sure. These two are addressed in the new commit.

**Comment by jiminha** [comment]
@cfgfung   can you merge your change to the latest baseline?

**Comment by cfgfung** [comment]
> @cfgfung can you merge your change to the latest baseline?

Hi,
I have applied a `git rebase`. It should be aligned with the latest baseline.

**Comment by libinta** [comment]
@cfgfung  can you rebase and provide the test result on latest?

**Comment by cfgfung** [comment]
> @cfgfung can you rebase and provide the test result on latest?

Hi,

I have applied rebase and ran the tests.

image

**Review by regisss — APPROVED** [review]

**Pull request merged**